src/main/java/io/vertx/ext/asyncsql/impl/AsyncSQLConnectionImpl.java

```diff
@@ -22,154 +22,314 @@
 import io.vertx.core.AsyncResult;
 import io.vertx.core.Future;
 import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
 import io.vertx.core.json.JsonArray;
 import io.vertx.ext.asyncsql.impl.pool.AsyncConnectionPool;
 import io.vertx.ext.sql.*;
 import scala.Option;
 import scala.concurrent.ExecutionContext;
 import scala.runtime.AbstractFunction1;
+import scala.util.Try;
 
 import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 /**
  * Implementation of {@link SQLConnection} using the {@link AsyncConnectionPool}.
  *
  * @author <a href="http://escoffier.me">Clement Escoffier</a>
+ * @author <a href="mailto:plopes@redhat.com">Paulo Lopes</a>
  */
-public class AsyncSQLConnectionImpl implements SQLConnection {
-
-  private final ExecutionContext executionContext;
-  private volatile boolean inTransaction = false;
-  private boolean inAutoCommit = true;
+abstract class AsyncSQLConnectionImpl implements SQLConnection {
 
+  private final Vertx vertx;
   private final Connection connection;
   private final AsyncConnectionPool pool;
+  private final ExecutionContext executionContext;
 
-  public AsyncSQLConnectionImpl(Connection connection, AsyncConnectionPool pool, ExecutionContext executionContext) {
+  // inTransaction can only be true when in !inAutoCommit, on switch the value should be false
+  // which means that the first statement to be executed should start a transaction
+  private final AtomicBoolean inTransaction = new AtomicBoolean();
+  // inAutoCommit true is the default, when set to false it means that the
+  // user is controlling the transitionally of the connection
```

Comment:
Yes, if a user does not specify an isolation level then the server default is used (previous behavior) when the user chooses the isolation level then the next transaction will have it.